### PR TITLE
Fix #34: Standardize issue kind

### LINKED ISSUE #34: standardize issue kind?
This relates to a long-standing discussion. We do not have a standard way to bucket issues in a SARIF log in a way that groups issues by required action. An issue that reports a tool internal error, for example, should be denoted (because it should not be suppressible, needs to be routed to a tool owner rather than code owner, etc.). The log file could also provide a mechanism to emit information that is supports other diagnostics (a 'note' in Clang terms). For compliance/auditing scenarios, it would be useful for issues to provide a 'pass' designation. Below is a proposal for grouping checks:

1) error
2) warning (we support these two severities due to prevalence in existing tools)
3) internal error (should not be suppressible)
4) log msg or 'note' (should have a 'relatesTo' field, so that groups of related messages can be preserved when sorting)
5) 'not applicable' (a useful designation when an analysis target is altogether invalid for analysis
6) 'pending' or 'observation' (this one's fuzzy but the idea is to emit output which can't be evaluated as pass/fail without post-processing/application of policy.

### PR DESCRIPTION
**Pull request by ghost**
We add a property `kind` to the `result` object, with values such as `pass`, `warning`, `error`, _etc_.

@michaelcfanning @nguerrera

### CODE REVIEW

File: document/CoreFormat.mdk [hunk 1]
```diff
@@ -361,12 +361,70 @@ A `result` object describes a single result produced by a static analysis tool.
...
+
+If present, the `kind` property shall have one of the following values, with the specified meanings:
+
+* `pass`: The rule specified by the `ruleId` property (&sect;[#result-ruleId]) was evaluated, and no problem was found.
+
+* `warning`: The rule specified by the `ruleId` property was evaluated, and a problem was found.
+
+* `error`: The rule specified by the `ruleId` property was evaluated, and a serious problem was found.
+
+* `pending`: The rule specified by the `ruleId` property was evaluated, but whether the result of the evaluation
```
Comment (michaelcfanning) [inline document/CoreFormat.mdk:383]: constant use of 'evaluated/evaluation' in this sentence. can we rework just a bit?

Comment (michaelcfanning) [inline document/CoreFormat.mdk]: Note that pending is intended to help decouple policy from results production in cases where it isn't desirable to tightly couple policy with the tool. Canonical example: a tool that reports an observation around the compiler toolchain used to build a binary coupled with a separate policy that requires a min-bar tool chain to 'pass'

File: document/CoreFormat.mdk [hunk 2]
```diff
@@ -361,12 +361,70 @@ A `result` object describes a single result produced by a static analysis tool.
...
+In some cases, team policy or corporate policy might determine if an analysis result constitutes a problem.
+~
+
+~ IsoExample
+A binary checker might produce a result stating that an analysis target was compiled with a particular version of the compiler.
+Subsequently, a post-processing tool might examine this result to determine whether the use of that version of the compiler
+complies with corporate policy.
+~
+
+* `notApplicable`: The rule specified by the `ruleId` property was not evaluated, because it does not apply to the
```
Comment (michaelcfanning) [inline document/CoreFormat.mdk:401]: You might provide an example here. For example, a security check might apply to 32-bit binaries only. Such a tool would produce a 'notApplicable' message when passed a 64-bit binary to inspect.

File: document/CoreFormat.mdk [hunk 3]
```diff
@@ -528,7 +586,7 @@ The `#pragma` warning construct in C# is such a syntactic construct.
 
 ### `fixes` property { #result-fixes }
 
-A `result` object may contain a property names `fixes`, whose value is a JSON array of `fix` objects (&sect;[#fix]).
+A `result` object may contain a property names `fixes`, whose value is a JSON array of zero or more `fix` objects (&sect;[#fix]).
```
Comment (michaelcfanning) [inline document/CoreFormat.mdk:594]: sorry, I know this keeps coming up, this language implies that it is ok to add an empty array, with no fixes, to the log file. how do we handle this elsewhere? i.e., do we permit other arrays both to be emitted entirely or to exist as an empty array to indicate there is no data associated with the property? do we permit this as a convenience for results production? it is arguably a slight inconvenience in results consumption. I'm not sure I have a strong feeling on this topic, trying to recall what we've said about it in the past and to ensure that we're consistent across the standard.

**Comment by michaelcfanning** [comment]
LGTM. see comments.

File: document/CoreFormat.mdk (hunk 1 above)
Comment (ghost) [inline document/CoreFormat.mdk:383]: @michaelcfanning  The first two need to be the same ("was evaluated ... the result of the evaluation"). I changed the last one to "processing".

File: document/CoreFormat.mdk (hunk 2 above)
Comment (ghost) [inline document/CoreFormat.mdk:401]: @michaelcfanning Done

File: document/CoreFormat.mdk (hunk 3 above)
Comment (ghost) [inline document/CoreFormat.mdk:594]: @michaelcfanning The only other place in the spec where we have an array of "zero or more" somethings is the `runLog.results` property, which is indeed mandatory.

Yes, having an array be optional is a slight inconvenience for consumers, who have to check for null before operating on the array.

OK, I'll change the property to mandatory ("shall").

Comment (ghost) [inline document/CoreFormat.mdk:594]: @michaelcfanning No, I changed my mind. Only one company asked for `fixes`. We can't require every tool to emit an empty array of `fixes` for every issue it finds.

**Pull request merged**